### PR TITLE
[VDO-5662, VDO-5663] Trim sysfs usage

### PR DESCRIPTION
**Pull request by raeburn**
Pull in three changes from Mike Snitzer, and adjust internal code accordingly.

Mike's changes remove our sysfs support, then re-add log_level as a simple numeric module parameter.
Changes from Mike's version:
 - conditionalize most code changes on VDO_UPSTREAM
 - remove files from manifest instead of deleting them
 - change name of now-non-static "log_level" variable in case we're linked into the kernel
 - add documentation of the one remaining sysfs field

Also updated the target version number to note the removal of an interface.

### CODE REVIEW

**Comment by bjohnsto** [comment]
I'm wondering about doing the removal this way. I understand we don't want to bother fixing up our tests that use sysfs so we're just removing by using the VDO_UPSTREAM flag, but I am a bit concerned that the dm-linux perl testing will fail once this PR gets pushed over there. Currently, we only do checkin tests I think, but that may be enough to fail some tests. Might it be better to bite the bullet and remove it for everything, including our internal stuff?

**Review by bjohnsto — COMMENTED** [review]
See the main comment I left in the PR

File: src/c++/vdo/base/dm-vdo-target.c [hunk 1]
```diff
@@ -1073,9 +1081,12 @@ static void vdo_io_hints(struct dm_target *ti, struct queue_limits *limits)
...
 	 *
-	 * The value is displayed in sysfs, and also used by dm-thin to determine whether to pass
-	 * down discards. The block layer splits large discards on this boundary when this is set.
+	 * The value is used by dm-thin to determine whether to pass down discards. The block layer
+	 * splits large discards on this boundary when this is set.
 	 */
+#ifndef VDO_UPSTREAM
+	/* The value is also displayed in sysfs.  */
+#endif
 	limits->max_discard_sectors =
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/dm-vdo-target.c:1090]: Do we really need this comment just for ourselves?

File: src/packaging/kpatch/MANIFEST.yaml [hunk 2]
```diff
@@ -60,6 +64,9 @@ tarballs:
             - VDO_UPSTREAM
           +postProcessor: ../github/removeInternal.sh
         src/c++/uds/kernelLinux/uds:
+          +excludes:
+            - uds-sysfs.c
+            - uds-sysfs.h
           +defines:
             - VDO_UPSTREAM
```
Comment (lorelei-sakai) [inline src/packaging/kpatch/MANIFEST.yaml:71]: Not wrong, but this will conflict with the indexer subdirectory work. Just to keep it in mind.

**Review by lorelei-sakai — COMMENTED** [review]
The primary issue is the heavy reliance on VDO_UPSTREAM for #ifdefs. Usually code is needed for testing, or for userspace, and we can be more specific. VDO_UPSTREAM exclusions are very hard to test in vdo-devel, but also hard to fix up after the fact if we discover they are incorrect.

**Review by lorelei-sakai — CHANGES_REQUESTED** [review]
Forgot to click the "request changes" button.

**Comment by lorelei-sakai** [comment]
> I'm wondering about doing the removal this way. I understand we don't want to bother fixing up our tests that use sysfs so we're just removing by using the VDO_UPSTREAM flag, but I am a bit concerned that the dm-linux perl testing will fail once this PR gets pushed over there. Currently, we only do checkin tests I think, but that may be enough to fail some tests. Might it be better to bite the bullet and remove it for everything, including our internal stuff?

We could potentially do that part as a follow-up PR. It's not clear to me how big those dependencies are, and it should all be internal only.

**Review by lorelei-sakai — CHANGES_REQUESTED** [review]
It looks tlike there are some rebasing issues, and if 105 goes in first there ill be more (sorry).

I think you forgot to update the packaging/github manifest, also.

File: src/packaging/src-dist/MANIFEST.yaml [hunk 3]
```diff
@@ -358,8 +354,6 @@ tarballs:
             - thread-registry.c
```
Comment (lorelei-sakai) [inline src/packaging/src-dist/MANIFEST.yaml:354]: Do perl tests pass in you remove the sysfs files from src-dist/MANIFEST.yaml? I though we used those rpms to run our perl tests. I rather though you'd have to exclude these files from the github and kpatch manifests explicitly.

File: src/c++/vdo/base/status-codes.c [hunk 4]
```diff
@@ -52,7 +52,9 @@ const struct error_info vdo_status_list[] = {
 	{ "VDO_BAD_NONCE", "Bad nonce" },
 	{ "VDO_JOURNAL_OVERFLOW", "Journal sequence number overflow" },
 	{ "VDO_INVALID_ADMIN_STATE", "Invalid operation for current state" },
+#if defined(VDO_INTERNAL) || defined(INTERNAL)
 	{ "VDO_CANT_ADD_SYSFS_NODE", "Failed to add sysfs node" },
+#endif
 };
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/status-codes.c:59]: This is the new rebase conflict I mentioned; I added an INTERNAL error code bock so this could just get added to that.

File: src/c++/vdo/base/dedupe.c [hunk 5]
```diff
@@ -3098,19 +3110,25 @@ int vdo_message_dedupe_index(struct hash_zones *zones, const char *name)
...
 int vdo_add_dedupe_index_sysfs(struct hash_zones *zones)
 {
 	int result = kobject_add(&zones->dedupe_directory,
 				 &zones->completion.vdo->vdo_directory, "dedupe");
 
 	if (result == 0) {
-		vdo_set_admin_state_code(&zones->state,
-					 VDO_ADMIN_STATE_NORMAL_OPERATION);
+		vdo_set_dedupe_state_normal(zones);
 	}
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/dedupe.c:3126]: Can we call vdo_set_dedupe_state_normal() unconditionally at some point in the VDO load process? It feels weird that the internal code calls this necessary thing at a different time than the upstream code.

File: src/c++/vdo/base/dm-vdo-target.c [hunk 6]
```diff
@@ -2387,11 +2397,16 @@ static void load_callback(struct vdo_completion *completion)
...
 		vdo_continue_completion(completion, vdo_initialize_kobjects(vdo));
 		return;
 
+#endif
 	case LOAD_PHASE_LOAD_DEPOT:
+#if !defined(VDO_INTERNAL) && !defined(INTERNAL)
+		vdo_set_dedupe_state_normal(vdo->hash_zones);
+#endif
 		if (vdo_is_read_only(vdo)) {
 			/*
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/dm-vdo-target.c:2411]: Here. Can we just call vdo_set_dedupe_state_normal() unconditionally here?

**Review by lorelei-sakai — CHANGES_REQUESTED** [review]
I'm not sure I have the brain for a full proper review this evening, but I added some comments at least. Github thinks you have a conflict in uds-sysfs.c, but I don't see what it is. I commented on the status-codes.[ch] conflicts, though.

I think I'm done with the large patches, sorry about the conflicts.

File: src/packaging/kpatch/MANIFEST.yaml [hunk 7]
```diff
@@ -124,6 +128,7 @@ tarballs:
             - thread-registry.c
             - thread-registry.h
             - thread-utils.c
```
[the remark this answers is not in the document]
Comment (raeburn) [reply src/packaging/kpatch/MANIFEST.yaml:126]: I think it was outdated, I've been working on another iteration with the latest pull from vdo-devel and fixing the manifests.

File: src/c++/vdo/base/dedupe.c [hunk 8]
```diff
@@ -2572,10 +2578,14 @@ void vdo_free_hash_zones(struct hash_zones *zones)
 		vdo_finish_dedupe_index(zones);
 
 	ratelimit_state_exit(&zones->ratelimiter);
+#if defined(VDO_INTERNAL) || defined(INTERNAL)
 	if (vdo_get_admin_state_code(&zones->state) == VDO_ADMIN_STATE_NEW)
 		uds_free(zones);
 	else
```
[the remark this answers is not in the document]
Comment (raeburn) [reply src/c++/vdo/base/dedupe.c:2580]: Yep, updating

File: src/c++/vdo/base/dm-vdo-target.c (hunk 6 above)
Comment (raeburn) [reply src/c++/vdo/base/dm-vdo-target.c:2411]: I think so, yes; trying it out...

File: src/c++/vdo/base/dm-vdo-target.c [hunk 9]
```diff
@@ -36,9 +36,11 @@
 #include "logger.h"
 #include "memory-alloc.h"
 #include "message-stats.h"
+#if defined(VDO_INTERNAL) || defined(INTERNAL)
 #ifdef __KERNEL__
 #include "pool-sysfs.h"
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/dm-vdo-target.c:41]: We could combine these #ifdef into one clasue but I suppose it would be a confusing clause.

Do we need more than __KERNEL__ and TEST_INTERNAL here?
 (Also the uds-sysfs.h include farther down).

File: src/c++/vdo/base/dm-vdo-target.c [hunk 10]
```diff
@@ -3084,7 +3097,9 @@ static int __init vdo_init(void)
 #ifdef __KERNEL__
 	/* Memory tracking must be initialized first for accurate accounting. */
 	vdo_memory_init();
+#ifdef VDO_INTERNAL
 	uds_init_sysfs();
+#endif
 
 	vdo_initialize_thread_device_registry();
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/dm-vdo-target.c:3104]: As long as this needs updating, I'd get rid of the blank line, too. The UDS/VDO distinction it was trying to highlight is effectively gone.

File: src/c++/uds/kernelLinux/uds/uds-sysfs.c [hunk 11]
```diff
@@ -241,94 +222,15 @@ static const struct kobj_type memory_object_type = {
 };
 
 #endif /* TEST_INTERNAL or VDO_INTERNAL */
```
Comment (lorelei-sakai) [inline src/c++/uds/kernelLinux/uds/uds-sysfs.c:224]: Are the changes in uds-syfs.c and sysfs.c due to the fact that the log_level functions just go away?

It might be less confusing to split those out into a separate commit since they're internal--only changes.

File: src/c++/vdo/base/vdo.h [hunk 12]
```diff
@@ -274,10 +278,12 @@ struct vdo {
...
 	struct volume_geometry geometry;
 
+#if defined(VDO_INTERNAL) || defined(INTERNAL)
 	/* For sysfs */
 	struct kobject vdo_directory;
 	struct kobject stats_directory;
 
+#endif
 	/* N blobs of context data for LZ4 code, one per CPU thread. */
 	char **compression_context;
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/vdo.h:288]: As long as you're fixing whitespace, can you shift the #ifdef for data_vio-count (line 291) to get rid of that double blank line, too?

File: src/c++/vdo/base/dm-vdo-target.c [hunk 13]
```diff
@@ -3141,6 +3141,9 @@ module_init(vdo_init);
 module_exit(vdo_exit);
 
 #ifdef __KERNEL__
+module_param_named(log_level, vdo_log_level, uint, 0644);
+MODULE_PARM_DESC(log_level, "Log-level for log messages");
+
```
Comment (lorelei-sakai) [inline src/c++/vdo/base/dm-vdo-target.c:3146]: "Log level" does not need a hyphen in it, can you fix?

**Review by lorelei-sakai — CHANGES_REQUESTED** [review]
Generally looks good. I noted a number of small things.
I like that you cleaned up a bunch of things that Mike overlooked in his rough cut.

Some of the internal only bits I think are better split out into a new internal-only commit.

Also in the commit message for the second commit, it's a bit of a run-on. Can you start a new sentence at "It can be read..."?

File: src/packaging/rpm/kvdo.spec [hunk 14]
```diff
@@ -31,7 +31,14 @@ deduplication, compression, and thin provisioning.
 %post
 set -x
 /usr/sbin/dkms --rpm_safe_upgrade add -m %{kmod_name} -v %{version}
-/usr/sbin/dkms --rpm_safe_upgrade build -m %{kmod_name} -v %{version}
+if ! /usr/sbin/dkms --rpm_safe_upgrade build -m %{kmod_name} -v %{version}; then
```
[the remark this answers is not in the document]
Comment (raeburn) [reply src/packaging/rpm/kvdo.spec:34]: Oops that was indeed meant to be separate. At least, if I can get confirmation it's actually doing what I intended, which I haven't so far...

File: src/c++/vdo/base/dedupe.c [hunk 15]
```diff
@@ -2565,10 +2573,14 @@ void vdo_free_hash_zones(struct hash_zones *zones)
 		vdo_finish_dedupe_index(zones);
 
 	ratelimit_state_exit(&zones->ratelimiter);
+#if defined(VDO_INTERNAL) || defined(INTERNAL)
 	if (vdo_get_admin_state_code(&zones->state) == VDO_ADMIN_STATE_NEW)
```
[the remark this answers is not in the document]
Comment (raeburn) [reply src/c++/vdo/base/dedupe.c:2577]: Yes, kobject_put calls a callback (type->release) if the refcount goes to 0; in the sysfs world that's how some of our objects get freed

File: src/c++/uds/kernelLinux/uds/uds-sysfs.c (hunk 11 above)
Comment (raeburn) [reply src/c++/uds/kernelLinux/uds/uds-sysfs.c:224]: Yes... I'll split those out

**Review by lorelei-sakai — APPROVED** [review]
This all looks good to me. I hope it passes tests.
